### ISSUE TITLE
All scripts should exist in a block including Discovery no loose scripts should be allowed.

### ISSUE THREAD

**Issue by simonsabin** [pester/Pester#1763]
## Summary of the feature request

The transition to Discovery and Run makes it vital that you understand what runs when. The transition from 4 makes it hard often to figure out why things aren't working, when you have loose script code that sets things up. Especially for test discovery.

If ALL loose scripts had to exist in a BeforeAll, BeforeEach BeforeDiscovery After etc then you couldn't run foul of finding a script not running when expected.

In forcing you to put code in a specifc named block means its super clear what is being done when.

Currently migrating from 4 your loose scripts that ran at runtime still run but only at discovery time. Forcing folks to put them in a BeforeAll, BeforeEach or Before Discovery ensures folks are making the choice. 

## How should it work? (optional)

On discovery if loose scripts are found then pester should error, or warn of the existence of loose scripts.

**Comment by nohwnd** [comment]
any suggestions how to do that? The only thing I can think is using AST / Regex on every script, which wastes time if the script is correct.

**Comment by simonsabin** [comment]
I don't know the code, do you control the running of the blocks or do they just run. I was hoping it was the former and thus you were aware of it being a block outside of a pester block.
AST is another option and maybe its not a runtime thing but a migration thing. Evaluate all blocks and then provide a choice, 1. make it a beforeAll, 2 make it a BeforeDiscovery or 3 both. 
Its one of the things I really like, but it took a few days to move a simple module from 4 -5 due to this and the change in the test generation.

**Comment by nohwnd** [comment]
I control when the scriptblocks that are provided by the user to any of the Pester functions will run. BUT I don't control the code that is running in them or anywhere else. If I did I would not need the functions. 

Example: 

```powershell
"a"
Describe "describe" { 
    "b"
    It "it" { 
        "c"
    }
}
``` 

In this case I control when the Describe scriptblock will execute (in discovery) and when the It scriptblock will execute. But I have no idea that "a" excuted, or that "b" executed - those are the common cases of the code placed outside of Pester leaf blocks. 
And further I also don't know that "c" executed, unless it throws.

**Comment by simonsabin** [comment]
Firstly due you think there is value. It would have to be AST I would have thought. Maybe its an option for debug. I'd put most of my unexpected behaviour issues in pester due to these scripts.

**Comment by nohwnd** [comment]
Yes there would be value in this. But it would be better as a migration analyzer that would help you migrate, instead of analyzing the file on every run. You pretty much need to report any code that is not in the leaf block (that is "a" and "b" from the above example.

**Comment by simonsabin** [comment]
Had a go and think this does a good pass. 
https://github.com/sabinio/sabin.io.psrules/blob/master/src/Rules.psm1

I'm sure there are a number of assumptions that are probably wrong.  It currently reports commands that are Pester that lie outside of an Before.., After.. or It block

**Comment by nohwnd** [comment]
Will have to look at this after 5.1 is released, if I seem to take too long, please ping me again :)

**Comment by simonsabin** [comment]
Thanks. I've enhanced it to provide a suggestion to wrap in a BeforeAll.

**Comment by nohwnd** [comment]
There's real value in the idea, and back then the right shape was a migration/analyzer rule rather than a per-run check. The landscape has shifted though: AI/LLM tooling now makes spotting and migrating loose top-level code (into BeforeAll/BeforeDiscovery/It) straightforward, which covers the original pain far more flexibly than a built-in warning would. Thanks @simonsabin for the prototype rules. Closing this as won't-do.
